flang/lib/Optimizer/CodeGen/ProcedurePointer.cpp

```diff
@@ -0,0 +1,302 @@
+/// Options to the procedure pointer pass.
+struct ProcedurePointerOptions {
+  // Lower the boxproc abstraction to function pointers and thunks where
+  // required.
+  bool useThunks = true;
+};
+
+/// This type converter rewrites all `!fir.boxproc<Func>` types to `Func` types.
+class BoxprocTypeRewriter : public mlir::TypeConverter {
+public:
+  using mlir::TypeConverter::convertType;
+
+  BoxprocTypeRewriter() {
+    addConversion([](mlir::Type ty) { return ty; });
+    addConversion([](BoxProcType boxproc) { return boxproc.getEleTy(); });
+    addConversion([&](mlir::TupleType tupTy) {
+      llvm::SmallVector<mlir::Type> memTys;
+      for (auto ty : tupTy.getTypes())
+        memTys.push_back(convertType(ty));
+      return mlir::TupleType::get(tupTy.getContext(), memTys);
+    });
+    addArgumentMaterialization(materializeProcedure);
+    addSourceMaterialization(materializeProcedure);
+    addTargetMaterialization(materializeProcedure);
+  }
+
+  static mlir::Value materializeProcedure(mlir::OpBuilder &builder,
+                                          BoxProcType type,
+                                          mlir::ValueRange inputs,
+                                          mlir::Location loc) {
+    assert(inputs.size() == 1);
+    return builder.create<ConvertOp>(loc, unwrapRefType(type.getEleTy()),
+                                     inputs[0]);
+  }
+};
+
+/// Rewrite all `fir.emboxproc` ops to either `fir.convert` or a thunk as
+/// required.
+class EmboxprocConversion : public mlir::OpRewritePattern<EmboxProcOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  EmboxprocConversion(mlir::MLIRContext *ctx) : OpRewritePattern(ctx) {}
+
+  mlir::LogicalResult
+  matchAndRewrite(EmboxProcOp embox,
+                  mlir::PatternRewriter &rewriter) const override {
+    mlir::Type toTy = embox.getType().cast<BoxProcType>().getEleTy();
+    if (embox.host()) {
+      // Create the thunk.
+      auto module = embox->getParentOfType<mlir::ModuleOp>();
+      FirOpBuilder builder(rewriter, getKindMapping(module));
+      auto loc = embox.getLoc();
+      mlir::Type i8Ty = builder.getI8Type();
+      mlir::Type i8Ptr = builder.getRefType(i8Ty);
+      mlir::Type buffTy = SequenceType::get({32}, i8Ty);
+      auto buffer = builder.create<AllocaOp>(loc, buffTy);
+      mlir::Value closure = builder.createConvert(loc, i8Ptr, embox.host());
+      mlir::Value tramp = builder.createConvert(loc, i8Ptr, buffer);
+      mlir::Value func = builder.createConvert(loc, i8Ptr, embox.func());
+      builder.create<fir::CallOp>(
+          loc, factory::getLlvmInitTrampoline(builder),
+          llvm::ArrayRef<mlir::Value>{tramp, func, closure});
+      auto adjustCall = builder.create<fir::CallOp>(
+          loc, factory::getLlvmAdjustTrampoline(builder),
+          llvm::ArrayRef<mlir::Value>{tramp});
+      rewriter.replaceOpWithNewOp<ConvertOp>(embox, toTy,
+                                             adjustCall.getResult(0));
```

Comment:
I think @klausler had in mind trampolines could be made in the runtime somehow to avoid making the stack executable. When we want to try that approach, do you think it would be pluggable here ?

Reply:
That's right.  Does this patch here imply a stack segment with execution privileges?  That would be a huge security problem.

Reply:
Yes, that's the idea for sure. Based on the target triple, this rewrite will have to take some different actions.